### ISSUE TITLE
InfluxDB delete with predicate works for _measurement, but not for specific _value

### ISSUE THREAD

**Issue by snenkov** [influxdata/influxdb#20399]
__Steps to reproduce:__
List the minimal actions needed to reproduce the behavior.

Here is the code I use (in python) to try and delete a specific `_value`:
```
def delete_data(data: str, org: str, bucket: str, token: str, db_url=INFLUX_URL) -> str:
    url = f'{db_url}/api/v2/delete?org={org}&bucket={bucket}'
    headers = {
        'Authorization': f'Token {token}',
        'Content-type': 'application/json'
    }
    response = requests.post(url, json=data, headers=headers)
    response.raise_for_status()
    return response

data = {
    "start": "2020-12-01T00:00:00Z",
    "stop": "2021-12-20T00:00:00Z",
    "predicate": "_field=\"my_field\""
}
r = delete_data(data=data, org=ORG, bucket='my_bucket', token=MY_TOKEN)def delete_data(data: str, org: str, bucket: str, token: str, db_url=INFLUX_URL) -> str:
    url = f'{db_url}/api/v2/delete?org={org}&bucket={bucket}'
    headers = {
        'Authorization': f'Token {token}',
        'Content-type': 'application/json'
    }
    response = requests.post(url, json=data, headers=headers)
    response.raise_for_status()
    return response

data = {
    "start": "2020-12-01T00:00:00Z",
    "stop": "2021-12-20T00:00:00Z",
    "predicate": "_field=\"my_field\""
}
r = delete_data(data=data, org=ORG, bucket='my_bucket', token=MY_TOKEN)
```

This does not delete the `_value`.
The same code, works just fine with `_measurement`

__Expected behavior:__
I expect the data to be deleted.

__Actual behavior:__
Nothing.

__Environment info:__

* System info: Linux 4.15.0-128-generic x86_64
* InfluxDB version: I am running it through docker: quay.io/influxdb/influxdb:v2.0.3 (when I do influxd version it does not show the docker instance)
* Other relevant environment details: Container runtime is up 7 days.

**Comment by danxmoran** [comment]
@snenkov can you provide some extra info to help me set up a realistic reproduction? Ideally:
* Example line-protocol inputs that would be affected by your delete call (ideally with multiple field=value pairs in their field-sets)
* Expected output of querying the bucket after executing the delete on the inputs

**Comment by snenkov** [comment]
Here is some example data: 
[2020-12-22-21-42_chronograf_data.csv.zip](https://github.com/influxdata/influxdb/files/5731733/2020-12-22-21-42_chronograf_data.csv.zip)

One the delete is executed, there should be no _field of `close_price` (so each row with this field should be gone)

**Comment by danxmoran** [comment]
I think this might be bigger problem with our InfluxQL compatibility. If I load your test data in and send a V1 query:
```
select * from price_volume_1_min where _field = "close_price"
```
I get 0 results.

**Comment by snenkov** [comment]
Thanks @danxmoran. This confirms the bug then.
So, what are the options to delete it now? Is there work which needs to be done on the backend, or some loophole that we can use for now?

**Comment by danxmoran** [comment]
I'm honestly not sure. I'll raise it internally, but I'm not sure how much attention it'll get until after the holiday season since so many people are on vacation.

**Comment by snenkov** [comment]
Okay cool. Thanks. Yeah now is kind of a laid-back time. :)
It’s important to have it fixed at some point. Otherwise we can’t really delete data :/

**Comment by danxmoran** [comment]
#6150 is the (ancient) parallel from the 1.x line

**Comment by snenkov** [comment]
I cannot believe this does not exist. It's quite crazy : / Thanks for linking to the #6150 . I hope it gets resolved soon.

**Comment by ski2day** [comment]
Thanks to Dan for pointing me here.

This is a major problem.  Not only is a broken feature documented as something that works, some people (such as myself) are totally blocked from the most basic database functionality; deleting data points in a time series.  I am blocked.

**Comment by MzazM** [comment]
Agreed. It is documented as working and it seems a necessary feature when working with time series.

**Comment by Aqualie** [comment]
Any update on this? Also recently just ran into this when attempting to onboard a new dataset and the value type was set incorrectly seems I now have to delete the entire measurement as I cannot do a _field="name" delete..

**Comment by Znubbis** [comment]
Still nothing on this after so long time?
It's like creating a file system and releasing it without a way to delete some of the files.

We need a way to clean up data that is no longer needed or in my case was added by misstake. A fundamental feature in any database I might say.

**Comment by mishop** [comment]
```
curl --request POST "influx:/api/v2/delete?org=Org&bucket=heating" --header 'Authorization: Token api' --header 'Content-Type: application/json' --data '{                                            
    "start": "2021-12-26T02:08:00Z",
    "stop": "2022-01-09T02:08:03Z",               
    "predicate": "_time=\"2022-01-06T17:05:47Z\" "            
  }'
```

Titme tag not working.

**Comment by dg-eparizzi** [comment]
I don't get it. I was able to delete a specific _field in InfluxDB Cloud but not on 2.0.7. Is this really NOT supported in the OSS version?
If so, please update documentation here: https://docs.influxdata.com/influxdb/v2.0/reference/syntax/delete-predicate/ which states that it cannot be used with `_time` or `_value` but it doesn't mention that it doesn't work with `_field`.

**Comment by DutchEllie** [comment]
This issue has been silent for so long, how come this has not even been implemented yet? Deleting data based on the common name seems like the most basic of features, doesn't it?

**Comment by 2opremio** [comment]
I could work around the issue by saving the measurement to a temporary measurement, deleting the original measurement, and then restoring the measurement from the copy.

**Comment by eni23** [comment]
Is there any update on this? This is really a bad bug and some of the things that bug me a lot in V2.0

**Comment by yozik04** [comment]
Still does not work in 2.6.1

**Comment by rwb196884** [comment]
Seems to be still fucked; I do
```
# influx delete --org mini31 --bucket solar --start 2020-01-01T00:00:00Z --stop 2023-02-19T00:00:00Z --predicate '_field="batt_level"'
# influx version
Influx CLI 2.6.1 (git: 61c5b4d) build_date: 2022-12-29T15:41:09Z
```
and nothing happens.

**Comment by sanderson** [comment]
InfluxDB OSS 2.x does not support deleting by field, as noted in the documentation: https://docs.influxdata.com/influxdb/v2.6/write-data/delete-data/#cannot-delete-data-by-field

**Comment by eni23** [comment]
> This is a limitation of the underlying storage engine used in 2.x.

Yes, everyone knows. Still, this is not a limitation, it is a bug from a users point of view. This is a really simple operation, which worked in 1.x and in 2.x it just fails silently.

I can recommend everyone to look for alternatives. Such ridiculos limitations together with the powerful, but weird-ass query language not suitable for non-techies makes influx 2.x just not considerable anymore.

**Comment by Subnum12** [comment]
> InfluxDB OSS 2.x does not support deleting by field, as noted in the documentation: https://docs.influxdata.com/influxdb/v2.6/write-data/delete-data/#cannot-delete-data-by-field

That's nice way to handle bugs, just change the specification instead of fixing the issue... (yes, sometimes bugs in design requires significant changes).....

**Comment by NTong97** [comment]
So this will never get addressed, huh??

**Comment by ski2day** [comment]
When looking at how to struture your data, the use of 'measurement' is very intuitive.  However, if you can never use 'measurement' to simplify data deletion, then why is it still here?  While tags are far less intuitive to use for structuring data, it may the be only solution to more easily clean up data easily.

This is the most frustrating aspect of influxDB.  The general outcome is that I don't bother cleaning things up as much as I should.  I am considering resturctuing to use tags instead of measurment as a result, but suspect that this comes at a cost in efficiency.  But, If I need to go this length to make deleting data simpler, then I should probably look at other options working with time series data given the solutions have been evolving.

**Comment by Roberto6969** [comment]
It would be fine if someone will fix this "bug" - yes, it is a bug if you can't perform such a simple task as delete points from field is. Deleting was implemented in earlier versions so I can't see any reason to implement sucha a basic function in latest versions. Don't throw dirt on an otherwise great product.

**Comment by prithvisathiya** [comment]
The fact that a core feature of deleting by field name was just silently removed is absolutely abysmal. Especially since it works on the Cloud version for 2.X. Whoever is making these decisions internally must be completely delusional

**Comment by philjb** [comment]
Hey @prithvisathiya - while I understand the frustrations, I'd like to help clarify some confusion.

InfluxDB 1.x oss and 2.x oss have never supported deleting by "field key" - a delete predicate that, for example, specifies `temp = 35`. So this ability was not removed in 2.x oss. 

InfluxDB 1.x and 2.x index tag names and tag values - not fields - and thus deleting by field is not available - Scott referenced this in [his comment](https://github.com/influxdata/influxdb/issues/20399#issuecomment-1435755691). Both [1.x oss](https://docs.influxdata.com/influxdb/v1/query_language/manage-database/#delete-series-with-delete) and [2.x oss](https://docs.influxdata.com/influxdb/v2/write-data/delete-data/) support deleting by a tag or measurement predicate or time range or dropping entire buckets or retention policies. 

Cloud 2 on TSM additionally supports deleting by field name (not value) e.g. `_field=temp` - [docs](https://docs.influxdata.com/influxdb/cloud/reference/syntax/delete-predicate/). This is available because Cloud 2 on TSM added indexing on field names, which also makes filtering on fields at query time faster at the expense of fields now contributing to the series key and thus series cardinality. 2.x OSS came out before Cloud 2 on TSM, although the development overlapped.  

I think @Roberto6969 made a typo in [the comment](https://github.com/influxdata/influxdb/issues/20399#issuecomment-2095367947) by leaving out a "not": "Deleting was [not] implemented in earlier versions so ..." I think this caused some of the confusion.

To add to the complexity, the default storage engine in Cloud 2 aka "serverless" is now iox/v3 if you go to [the signup](https://cloud2.influxdata.com/signup), which does [not support delete predicates](https://docs.influxdata.com/influxdb3/cloud-serverless/reference/syntax/delete-predicate/) - deleting by retention or [dropping buckets](https://docs.influxdata.com/influxdb3/cloud-serverless/reference/cli/influx/bucket/delete) is the way to remove data. I know engineering is working to add row level deletes (equivalent to delete predicates) to iox/v3 storage engine this year. Given the timeline, I understand if you need the feature sooner. 

You can get a Cloud 2 account with a TSM storage backend by emailing support (support@influxdata.com) - but I believe it has to be a pay-as-you-go account. 

The newest OSS offering is called ["Core"](https://www.influxdata.com/products/influxdb-core/), which is available and with new features in development. Core [supports](https://docs.influxdata.com/influxdb3/core/reference/cli/influxdb3/delete) database and table deletes like serverless. The next release (coming this week or next I believe) [enhances](https://github.com/influxdata/influxdb/pull/26553) these delete capabilities (available now if built from source).

**Comment by prithvisathiya** [comment]
@philjb 
thanks for the explanation and clarification. I am currently already on the Cloud Managed version, but am currently migrating from Influx Cloud to self-hosted due to internal non-adjustable throttle limits when users query the DB (see the 25m query exceeded in past 30s error). Unfortunately their dedicated hosted version only supports Influx V3, which includes lots of changes compared to V2

But the OSS includes just a lot of frustrating features (or limitations). We already discussed the inabiltity to delete by `field` , but I just realized, even deleting by `tags` don't work as expected! Deleting the tag values, do delete the data and all the fields associated with it, BUT it still retains the schema of the field. So if I want to delete the field in order to change the data type, I not only _cannot_ delete by field, but i also cannot delete by tag. So I essentially have to delete the entire measurement....

I understand that there are technical limitations underneath how Influx OSS is built, but at the end of the day, the user is unable to do perform basic operations available in other DBs

**Comment by philjb** [comment]
@prithvisathiya - For OSS, I have tactical suggestions but it does require another step. If you [delete by tag](https://github.com/influxdata/influxdb/issues/25471#issuecomment-2419938222), you can wait for a new shard to be created which won't keep the schema of the previous shard. If you're doing infinite bucket retention, I think the default shard duration is 7 days. Alternatively, you can [rebuild the tsi index](https://docs.influxdata.com/influxdb/v2/admin/internals/tsi/rebuild-index/) after the delete. 

I believe the [Global limit](https://docs.influxdata.com/influxdb/cloud/account-management/limits/#global-limits) of query time can be adjusted for annual contract customers of a certain size and pending engineering review; although I know they are reluctant to adjust this limit. Or perhaps your queries could be optimized to execute faster?

While there are differences in V2 and V3, the experience of V3 on "Cloud Serverless" can be similar to Cloud2 on TSM. Cloud2 is TSM backed - same as OSS 2.x. Cloud Serverless is V3 (parquet + DataFusion) backed and has different [global limits](https://docs.influxdata.com/influxdb3/cloud-serverless/admin/billing/limits/). While V3 prefers your queries are InfluxQL or SQL, there is a [deprecated api](https://docs.influxdata.com/influxdb3/cloud-serverless/api/v2/#operation/PostQuery) for flux queries in V3. It is deprecated because sometimes a flux query optimized for TSM runs slower on V3 Serverless.
